src/Sema.zig

```diff
@@ -29948,8 +29948,20 @@ fn coerceArrayLike(
 
     const dest_elem_ty = dest_ty.childType(mod);
     const inst_elem_ty = inst_ty.childType(mod);
-    const in_memory_result = try sema.coerceInMemoryAllowed(block, dest_elem_ty, inst_elem_ty, false, target, dest_ty_src, inst_src);
-    if (in_memory_result == .ok) {
+    const can_bitcast: bool = can_bitcast: {
+        const in_memory_result = try sema.coerceInMemoryAllowed(block, dest_elem_ty, inst_elem_ty, false, target, dest_ty_src, inst_src);
+        if (in_memory_result != .ok) break :can_bitcast false;
+
+        if ((dest_ty.zigTypeTag(mod) == .Vector and inst_ty.zigTypeTag(mod) == .Array) or (dest_ty.zigTypeTag(mod) == .Array and inst_ty.zigTypeTag(mod) == .Vector)) {
+            // memory layout of @Vector(N, iM) is the same as i(N*M), the padding is not packed like in [N]iM.
+            // but if N is a power of two, there's no padding.
+            const bit_size = dest_elem_ty.bitSize(mod);
+            break :can_bitcast bit_size >= 8 and std.math.isPowerOfTwo(bit_size);
+        } else {
+            break :can_bitcast true;
+        }
```

Comment:
I think this logic belongs in `coerceInMemoryAllowed`

Reply:
mmm the call to coerceInMemoryAllowed() is about the inner elements.
the test is about how the elements are layed out in memory by the vector/array.
Are you suggesting somehow merging coerceArrayLike() into coerceInMemoryAllowed()?